omegaconf/dictconfig.py

```diff
@@ -270,35 +270,33 @@ def _validate_and_normalize_key(self, key: Any) -> DictKeyType:
     def _s_validate_and_normalize_key(self, key_type: Any, key: Any) -> DictKeyType:
         if key_type is Any:
             for t in DictKeyType.__args__:  # type: ignore
-                try:
-                    return self._s_validate_and_normalize_key(key_type=t, key=key)
-                except KeyValidationError:
-                    pass
+                if isinstance(key, t):
+                    return key  # type: ignore
             raise KeyValidationError("Incompatible key type '$KEY_TYPE'")
-        elif key_type == str:
-            if not isinstance(key, str):
-                raise KeyValidationError(
-                    f"Key $KEY ($KEY_TYPE) is incompatible with ({key_type.__name__})"
-                )
-
-            return key
-        elif key_type == int:
-            if not isinstance(key, int):
+        elif key_type == bool and (key == 0 or key == 1):
+            # Python treats True as 1 and False as 0 when used as dict keys
+            #   assert hash(0) == hash(False)
+            #   assert hash(1) == hash(True)
+            return key  # type: ignore
+        elif key_type in (str, int, float, bool):  # primitive type
+            if not isinstance(key, key_type):
                 raise KeyValidationError(
                     f"Key $KEY ($KEY_TYPE) is incompatible with ({key_type.__name__})"
                 )
 
-            return key
+            return key  # type: ignore
         elif issubclass(key_type, Enum):
             try:
                 ret = EnumNode.validate_and_convert_to_enum(key_type, key)
-                assert ret is not None
-                return ret
+                if ret is None:
+                    raise ValidationError
             except ValidationError:
                 valid = ", ".join([x for x in key_type.__members__.keys()])
                 raise KeyValidationError(
                     f"Key '$KEY' is incompatible with the enum type '{key_type.__name__}', valid: [{valid}]"
                 )
+            else:
+                return ret
```

Comment:
I don't like this.
can you just put the return in the try?

Reply:
Sure. Done in https://github.com/omry/omegaconf/pull/484/commits/2fec127369cb14249758de30f1d5f1b0764108ba